lib/emscripten/src/invoke.rs

```diff
@@ -0,0 +1,217 @@
+use wasmer_runtime_core::backend::CompilerConfig;
+use wasmer_runtime_core::{compile_with_config, Module};
+
+struct Sig {
+    pub args: Vec<InvokeArgType>,
+    pub rets: Vec<InvokeArgType>,
+}
+
+pub fn create_invoke_module(invoke_function_names: &Vec<String>) -> Module {
+    let wat = create_wat(invoke_function_names);
+    let wasm_binary = wabt::wat2wasm(wat.as_bytes()).unwrap();
+    let compiler = wasmer_clif_backend::CraneliftCompiler::new();
+    let compiler_config = CompilerConfig { symbol_map: None };
+    let module = compile_with_config(&wasm_binary[..], &compiler, compiler_config);
+    module.unwrap()
+}
+
+fn create_wat(invoke_function_names: &Vec<String>) -> String {
+    let types_and_funcs_wat = invoke_function_names
+        .iter()
+        .map(|s| &**s) // &String -> &str
+        .map(invoke_name_to_signature)
+        .filter_map(|r| r.ok())
+        .map(|(func_name, sig)| {
+            let type_wat = sig_to_type_wat(func_name, &sig);
+            let func_wat = sig_to_func_wat(func_name, &sig);
+            format!("{}\n{}", type_wat, func_wat)
+        })
+        .collect::<Vec<_>>()
+        .join("\n");
+    format!(
+        "(module (table {} anyfunc)\n {})",
+        invoke_function_names.len(),
+        types_and_funcs_wat
+    )
+}
+
+fn invoke_name_to_signature(invoke_name: &str) -> Result<(&str, Sig), InvokeError> {
+    let mut args: Vec<InvokeArgType> = vec![];
+    let mut rets: Vec<InvokeArgType> = vec![];
+    let chars = invoke_name.chars();
+    // skip 'invoke_'
+    let mut chars = chars.skip(7);
+    // 'if next is a 'v' then this does not return
+    match chars.next() {
+        Some('v') => {}
+        Some(c) => match InvokeArgType::from_char(c) {
+            Ok(i) => rets.push(i),
+            Err(_) => return Err(InvokeError::InvalidInvokeFunctionParameterCharacter),
+        },
+        None => {}
+    };
+    // continue iterating
+    for c in chars {
+        match InvokeArgType::from_char(c) {
+            Ok(i) => args.push(i),
+            Err(_) => return Err(InvokeError::InvalidInvokeFunctionParameterCharacter),
+        }
+    }
+
+    let sig = Sig { args, rets };
+    Ok((invoke_name, sig))
+}
+
+fn sig_to_type_wat(func_name: &str, sig: &Sig) -> String {
+    let x = sig.args.iter().map(|i| i.to_type_str()).collect::<Vec<_>>();
+    let params_string = match &x[..] {
+        [] => "".to_string(),
+        _args => format!(" (param {})", x.join(" ")),
+    };
+    let result_string = match sig.rets.get(0) {
+        Some(ret) => format!(" (result {})", ret.to_type_str()),
+        None => "".to_string(),
+    };
+    format!(
+        "(type $emscripten_{} (func{}{}))",
+        func_name, params_string, result_string
+    )
+}
+
+fn sig_to_func_wat(func_name: &str, sig: &Sig) -> String {
```

Comment:
doc comments on some of the larger functions would be nice; brief explanation of what the `func wat` will look like would be good I think